File: sol/1139.cpp
struct Solution {
    int largest1BorderedSquare(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid.front().size();
        vector<vector<int>> vSeg(n, vector<int>(m, 0)), hSeg(n, vector<int>(m, 0));
        for (int i = 0; i < n; ++i){
            hSeg[i][0] = grid[i][0];
            for (int j = 1; j < m; ++j){
                if (grid[i][j] == 1){
                    hSeg[i][j] = hSeg[i][j - 1] + 1;
                }
            }
        }
        for (int j = 0; j < m; ++j){
            vSeg[0][j] = grid[0][j];
            for (int i = 1; i < n; ++i){
                if (grid[i][j] == 1){
                    vSeg[i][j] = vSeg[i - 1][j] + 1;
                }
            }
        }
        int res = 0;
        for (int d = n + m - 2; d >= 0; --d){
            if (d / 2 + 1 < res){
                break;
            }
            for (int i = min(n - 1, d); i >= max(0, d - m + 1); --i){
                int j = d - i;
                if (min(i, j) + 1 < res){
                    continue;
                }
                for (int l = min(hSeg[i][j], vSeg[i][j]); l >= 1; --l){
                    if (min(hSeg[i - l + 1][j], vSeg[i][j - l + 1]) >= l){
                        res = max(res, l);
                        break;
                    }
                }
            }
        }
        return res * res;
    }
};

File: sol/1139/1139.py
class Solution:
    def largest1BorderedSquare(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        vSegLen = [[0] * m for _ in range(n)]
        hSegLen = [[0] * m for _ in range(n)]
        for i in range(n):
            hSegLen[i][0] = grid[i][0]
            for j in range(1, m):
                if grid[i][j] != 0:
                    hSegLen[i][j] = hSegLen[i][j - 1] + 1
        for j in range(m):
            vSegLen[0][j] = grid[0][j]
            for i in range(1, n):
                if grid[i][j] != 0:
                    vSegLen[i][j] = vSegLen[i - 1][j] + 1
        res = 0
        for d in range(n + m - 2, -1, -1):
            for i in range(min(n - 1, d), max(-1, d - m), -1):
                j = d - i
                if min(i, j) + 1 < res:
                    continue
                for l in range(min(hSegLen[i][j], vSegLen[i][j]), 0, -1):
                    if min(hSegLen[i - l + 1][j], vSegLen[i][j - l + 1]) >= l:
                        res = max(res, l)
                        break
        return res ** 2
